### ISSUE TITLE
3.12.7: gh-113993 backported without subsequent fixes, causing `python3: Objects/unicodeobject.c:15018: intern_common: Assertion `_Py_IsImmortal(r)' failed.`

### ISSUE THREAD

**Issue by mgorny** [python/cpython#124887]
# Crash report

### What happened?

The 49f6beb56a64dd0b688e8c728928a38eb6d78fef ("Make interned strings mortal") backport introduced ABI-breaking assertions into 3.12.7. As a result, Rust stable ABI packages are now crashing on assertions, e.g. from `uv`'s `cargo test`:

```
          5 │+python3: Objects/unicodeobject.c:15018: intern_common: Assertion `_Py_IsImmortal(r)' failed.
          6 │+python3: Objects/unicodeobject.c:15018: intern_common: Assertion `_Py_IsImmortal(r)' failed.
```

Backporting 281ffb60cc194c4eefee3e112888251abd1a749a seems to fix the immediate issue but I haven't checked for more regressions.

### CPython versions tested on:

3.12

### Operating systems tested on:

Linux

### Output from running 'python -VV' on the command line:

Python 3.12.7 (main, Oct  1 2024, 21:49:30) [GCC 14.2.1 20240921]

**Comment by ZeroIntensity** [comment]
Ah, this problem again. `assert(_Py_IsImmortal(...))` breaks stable ABI extensions -- we're supposed to use `assert(_Py_IsImmortalLoose(...))`. IIRC, that PR took a while to backport, so there's a chance that it predated `_Py_IsImmortalLoose`. cc @encukou

**Comment by encukou** [comment]
I missed this in the big backport. My apologies :(

**Comment by encukou** [comment]
I've merged the backport: #124938

**Comment by mgorny** [comment]
Thank you!
